<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Quickstart · BioSequenceMappings.jl</title><meta name="title" content="Quickstart · BioSequenceMappings.jl"/><meta property="og:title" content="Quickstart · BioSequenceMappings.jl"/><meta property="twitter:title" content="Quickstart · BioSequenceMappings.jl"/><meta name="description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:description" content="Documentation for BioSequenceMappings.jl."/><meta property="twitter:description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/"/><meta property="twitter:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/"/><link rel="canonical" href="https://pierrebarrat.github.io/BioSequenceMappings.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>BioSequenceMappings.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Quickstart</a><ul class="internal"><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Usage"><span>Usage</span></a></li></ul></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="alphabets/">Alphabets</a></li><li><a class="tocitem" href="alignments/">Alignments</a></li><li><a class="tocitem" href="utilities/">Utilities</a></li></ul></li><li><a class="tocitem" href="reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Quickstart</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Quickstart</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/master/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="BioSequenceMappings"><a class="docs-heading-anchor" href="#BioSequenceMappings">BioSequenceMappings</a><a id="BioSequenceMappings-1"></a><a class="docs-heading-anchor-permalink" href="#BioSequenceMappings" title="Permalink"></a></h1><p>The aim of the package is to facilitate the task of converting biological sequences (nucleotides, amino acids) to integers or to onehot representation.  It also provides simple function to manipulate alignments or to compute simple statistics. </p><p><em>Note</em>: I do not frequently use the onehot format, so for now it is not documented / not well tested. I&#39;ll develop it more when I get the time. </p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><p>From the Julia REPL: </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using Pkg</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; Pkg.add(&quot;https://github.com/PierreBarrat/BioSequenceMappings.jl&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: `https://github.com/PierreBarrat/BioSequenceMappings.jl` is not a valid package name. Perhaps you meant `https://github.com/PierreBarrat/BioSequenceMappings`
The argument appears to be a URL or path, perhaps you meant `Pkg.add(url=&quot;...&quot;)` or `Pkg.add(path=&quot;...&quot;)`.</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; using BioSequenceMappings</code><code class="nohighlight hljs ansi" style="display:block;"></code></pre><h2 id="Usage"><a class="docs-heading-anchor" href="#Usage">Usage</a><a id="Usage-1"></a><a class="docs-heading-anchor-permalink" href="#Usage" title="Permalink"></a></h2><h3 id="Filter-sequences-by-Hamming-distance"><a class="docs-heading-anchor" href="#Filter-sequences-by-Hamming-distance">Filter sequences by Hamming distance</a><a id="Filter-sequences-by-Hamming-distance-1"></a><a class="docs-heading-anchor-permalink" href="#Filter-sequences-by-Hamming-distance" title="Permalink"></a></h3><p>Load an alignment, find all sequences with a Hamming distance smaller than 66% to the first one, create a new alignment object from them and save it to a file. </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; using BioSequenceMappings</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A = read_fasta(&quot;../../example/PF00014.fasta&quot;);</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; size(A) # 100 sequences of length 53</code><code class="nohighlight hljs ansi" style="display:block;">(53, 100)</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; s0 = A[1];</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; s0&#39; # transpose for legibility</code><code class="nohighlight hljs ansi" style="display:block;">1×53 adjoint(view(::Matrix{Int64}, :, 1)) with eltype Int64:
 1  3  10  10  14  14  17  5  7  19  3  …  17  15  15  4  3  9  17  18  3  1</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; indices = findall(s -&gt; hamming(s, s0; normalize=true) &lt; 0.66, A)</code><code class="nohighlight hljs ansi" style="display:block;">5-element Vector{Int64}:
  1
 33
 64
 83
 86</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; B = subsample(A, indices)</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=5 sequences of length L=53 - Shown as `MxL` matrix
5×53 adjoint(::Matrix{Int64}) with eltype Int64:
 1  3  10  10  14  14  17   5  7  19  …  17  15  15  4  3   9  17  18  3  1
 1  3   6  11  14  11  17  10  7   6     17  11  13  5  3  16  11   2  3  1
 6  3  11   5  14  14  21  18  7  16     17   5   4  4  3  12  16  18  3  1
 1  3  11  11  15  19   4   5  7  14     17  21  15  5  3  15  10  18  3  1
 1  3  16   5  17  10  13   5  7  14     17  10  17  5  3  11  17  18  3  1</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; write(&quot;PF00014_subsample.fasta&quot;, B)</code><code class="nohighlight hljs ansi" style="display:block;"></code></pre><h3 id="Remove-columns-with-too-many-gaps"><a class="docs-heading-anchor" href="#Remove-columns-with-too-many-gaps">Remove columns with too many gaps</a><a id="Remove-columns-with-too-many-gaps-1"></a><a class="docs-heading-anchor-permalink" href="#Remove-columns-with-too-many-gaps" title="Permalink"></a></h3><p>Load an alignment, find columns with more than 5% gaps and remove them, then create a new alignment object. </p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A = read_fasta(&quot;../../example/PF00014.fasta&quot;); # uses the default alphabet `aa_alphabet`</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; gap_digit = A.alphabet(&#39;-&#39;)</code><code class="nohighlight hljs ansi" style="display:block;">1</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; f1 = site_specific_frequencies(A)</code><code class="nohighlight hljs ansi" style="display:block;">21×53 Matrix{Float64}:
 0.41  0.05  0.04  0.03  0.02  0.01  …  0.01  0.01  0.01  0.02  0.02  0.62
 0.1   0.0   0.06  0.07  0.06  0.13     0.0   0.0   0.16  0.14  0.0   0.06
 0.0   0.95  0.0   0.0   0.0   0.0      0.99  0.0   0.0   0.0   0.98  0.0
 0.0   0.0   0.05  0.03  0.02  0.0      0.0   0.02  0.06  0.01  0.0   0.0
 0.0   0.0   0.07  0.09  0.05  0.02     0.0   0.28  0.15  0.01  0.0   0.07
 0.08  0.0   0.03  0.02  0.0   0.03  …  0.0   0.0   0.0   0.02  0.0   0.0
 0.0   0.0   0.01  0.02  0.01  0.0      0.0   0.01  0.01  0.0   0.0   0.0
 0.0   0.0   0.03  0.02  0.0   0.01     0.0   0.0   0.01  0.03  0.0   0.0
 0.13  0.0   0.02  0.01  0.0   0.06     0.0   0.05  0.02  0.06  0.0   0.0
 0.0   0.0   0.05  0.02  0.03  0.15     0.0   0.12  0.11  0.08  0.0   0.11
 ⋮                             ⋮     ⋱                    ⋮
 0.0   0.0   0.07  0.0   0.02  0.01     0.0   0.02  0.08  0.02  0.0   0.0
 0.03  0.0   0.0   0.0   0.62  0.22     0.0   0.0   0.0   0.0   0.0   0.0
 0.0   0.0   0.03  0.09  0.03  0.02     0.0   0.13  0.05  0.01  0.0   0.05
 0.04  0.0   0.07  0.0   0.03  0.07  …  0.0   0.06  0.13  0.05  0.0   0.08
 0.0   0.0   0.16  0.02  0.06  0.05     0.0   0.03  0.12  0.12  0.0   0.0
 0.02  0.0   0.04  0.01  0.0   0.01     0.0   0.02  0.01  0.23  0.0   0.0
 0.16  0.0   0.04  0.04  0.01  0.03     0.0   0.01  0.02  0.06  0.0   0.0
 0.0   0.0   0.0   0.0   0.0   0.0      0.0   0.01  0.0   0.0   0.0   0.0
 0.0   0.0   0.02  0.02  0.0   0.01  …  0.0   0.0   0.02  0.09  0.0   0.0</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; non_gapped_colums = findall(j -&gt; f1[gap_digit, j] &lt;= 0.05, 1:size(f1, 2))&#39; # transpose for legibility</code><code class="nohighlight hljs ansi" style="display:block;">1×51 adjoint(::Vector{Int64}) with eltype Int64:
 2  3  4  5  6  7  8  9  10  11  12  …  44  45  46  47  48  49  50  51  52</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; B = Alignment(A.data[non_gapped_colums&#39;, :], A.alphabet; A.names) # sequences are stored as columns</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=100 sequences of length L=51 - Shown as `MxL` matrix
100×51 adjoint(::Matrix{Int64}) with eltype Int64:
 3  10  10  14  14  17   5   7  19  3  …  17  15  15   4  3   9  17  18  3
 3  15  15  14  11   8   8   7  13  3     18   6   5   4  1   1   1   1  1
 3   8  11  14  16   5   5   7   5  3     18  16  10  15  3  16  10   9  3
 3  17  11  17  14  17  14   7  18  3     17  16   2   5  3   5  13  21  3
 3   2   6  14  12   5  10   7  14  3     16  10   5  10  3   5  10   6  3
 3  11  15  17  10   4   9   7  18  3  …  17   5   5   2  3   5  16   2  3
 3  13  11  14   9  19  16   7  14  3     17   5  10   5  3  10   5  21  3
 3  17  11  14   2  19  15   7  14  3     17  16   5   5  3   5   5  17  3
 3   2  11  10   2   4   4   7  14  3     17  11   5   5  3  10   5  10  3
 3  11   2   5  19   4  11  16  19  3     18  21   5   5  3  10   2  17  3
 ⋮                   ⋮                 ⋱               ⋮                 ⋮
 3  12   5  10  14   5  10   7  14  3     18  17  18   5  3  17  10  18  3
 3   7  11  14  10   5  18   7  14  3     17   5  17   5  3  10   5  18  3
 3  13  11  14  17  19  19   7   9  3     18  10   5  17  3   5  16  15  3
 3  17   6  14  14   4  20   7  17  3     18   6   5   9  3  15  10   2  3
 3   5  11  14  19  13   8   7  16  3  …  17  21   5   5  3   9   7  12  3
 3  10  11  16  11   4  15   7  17  3     18   5   5   5  3  10  10  18  3
 3   5  15  14  12   4   2   7  14  3     18   5   2  17  3   7  21   8  3
 3  21  19   7  16  13   5   7  21  3     18  15   5  17  3   5   2  16  3
 3  12  11  14  10   4  18   7  18  3     17  15   5   2  3   5   2  19  3</code></pre></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="alphabets/">Alphabets »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.7.0 on <span class="colophon-date" title="Monday 30 September 2024 16:15">Monday 30 September 2024</span>. Using Julia version 1.10.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
