<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · BioSequenceMappings.jl</title><meta name="title" content="Home · BioSequenceMappings.jl"/><meta property="og:title" content="Home · BioSequenceMappings.jl"/><meta property="twitter:title" content="Home · BioSequenceMappings.jl"/><meta name="description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:description" content="Documentation for BioSequenceMappings.jl."/><meta property="twitter:description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:url" content="https://PierreBarrat.github.io/BioSequenceMappings.jl/"/><meta property="twitter:url" content="https://PierreBarrat.github.io/BioSequenceMappings.jl/"/><link rel="canonical" href="https://PierreBarrat.github.io/BioSequenceMappings.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>BioSequenceMappings.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/master/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="BioSequenceMappings"><a class="docs-heading-anchor" href="#BioSequenceMappings">BioSequenceMappings</a><a id="BioSequenceMappings-1"></a><a class="docs-heading-anchor-permalink" href="#BioSequenceMappings" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/PierreBarrat/BioSequenceMappings.jl">BioSequenceMappings</a>.</p><ul><li><a href="#BioSequenceMappings.Alignment-Union{Tuple{T}, Tuple{AbstractMatrix{T}, Alphabet{T}}} where T"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#BioSequenceMappings.Alignment-Union{Tuple{AbstractMatrix{T}}, Tuple{T}} where T&lt;:Integer"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#BioSequenceMappings.Alignment"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#BioSequenceMappings.Alphabet"><code>BioSequenceMappings.Alphabet</code></a></li><li><a href="#BioSequenceMappings.compute_mapping-Union{Tuple{AbstractString}, Tuple{T}, Tuple{AbstractString, Type{T}}} where T&lt;:Integer"><code>BioSequenceMappings.compute_mapping</code></a></li><li><a href="#BioSequenceMappings.compute_weights"><code>BioSequenceMappings.compute_weights</code></a></li><li><a href="#BioSequenceMappings.compute_weights!-Tuple{Any, Any}"><code>BioSequenceMappings.compute_weights!</code></a></li><li><a href="#BioSequenceMappings.default_alphabet-Union{Tuple{Integer}, Tuple{T}, Tuple{Integer, Type{T}}} where T&lt;:Integer"><code>BioSequenceMappings.default_alphabet</code></a></li><li><a href="#BioSequenceMappings.pairwise_hamming-Tuple{Alignment, Alignment}"><code>BioSequenceMappings.pairwise_hamming</code></a></li><li><a href="#BioSequenceMappings.pairwise_hamming-Tuple{Alignment}"><code>BioSequenceMappings.pairwise_hamming</code></a></li><li><a href="#BioSequenceMappings.read_fasta-Tuple{AbstractString}"><code>BioSequenceMappings.read_fasta</code></a></li><li><a href="#BioSequenceMappings.site_specific_frequencies-Tuple{Any, Vararg{Any}}"><code>BioSequenceMappings.site_specific_frequencies</code></a></li><li><a href="#BioSequenceMappings.subsample-Tuple{BioSequenceMappings.AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample</code></a></li><li><a href="#BioSequenceMappings.subsample_random-Tuple{BioSequenceMappings.AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample_random</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alignment" href="#BioSequenceMappings.Alignment"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">mutable struct Alignment{T} where T &lt;: Integer</code></pre><pre><code class="nohighlight hljs">    dat::Matrix{T}
    alphabet::Union{Nothing, Alphabet{T}}
    weights::Vector{Float64} = ones(size(dat,1))/size(dat,1) # phylogenetic weights of sequences
    names::Vector{String} = fill(&quot;&quot;, size(dat, 1))</code></pre><p>Biological sequences as vectors of type <code>T&lt;:Integer</code>. <code>dat</code> stores sequences in <em>columns</em>: <code>size(dat)</code> returns a tuple <code>(L, M)</code> with <code>L</code> the length and <code>M</code> the number of sequences.</p><p><code>alphabet</code> represents the mapping between integers in <code>dat</code> and biological symbols (nucleotides, amino acids...). If <code>nothing</code>, the alignment cannot be mapped to biological sequences.</p><p><code>weights</code> represent phylogenetic weights, and are initialized to <code>1/M</code>. They must sum to 1. <code>names</code> are the label of sequences, and are expected to be in the same order as the columns of <code>dat</code>. They do not have to be unique, and can be ignored</p><p><strong><strong>Important</strong>: When built from a matrix, will <em>transpose</em> the input; if <code>size(dat) = (M, L)</code>,</strong></p><p><strong><code>X=Alignment(dat, alphabet)</code> will return an object with <code>size(X.dat) = (L, M)</code>.</strong></p><p><strong>In other words, assumes that the input matrix has sequences as rows.</strong></p><p><strong>Methods</strong></p><ul><li><code>getindex(X::Alignment, i)</code> returns a matrix/vector <code>X.dat[:, i]</code>.</li><li><code>for s in X::Alignment</code> iterates over sequences.</li><li><code>eachsequence(X::Alignment)</code> returns an iterator over sequences (<code>Vector{Int}</code>).</li><li><code>eachsequence_weighted(X::Alignment)</code> returns an iterator over sequences and weights as tuples</li><li><code>subaln(X::Alignment, idx)</code> constructs the subaln defined by index <code>idx</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alignment.jl#L7-L39">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alignment-Union{Tuple{AbstractMatrix{T}}, Tuple{T}} where T&lt;:Integer" href="#BioSequenceMappings.Alignment-Union{Tuple{AbstractMatrix{T}}, Tuple{T}} where T&lt;:Integer"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Alignment(data::AbstractMatrix{T}; alphabet = :auto, kwargs...)</code></pre><p>Keyword argument <code>alphabet</code> can be <code>:auto</code>, <code>:none</code> or <code>nothing</code>, or an input to <code>Alphabet</code>. Other keyword arguments are passed to the default constructor of <code>Alignment</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alignment.jl#L127-L132">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alignment-Union{Tuple{T}, Tuple{AbstractMatrix{T}, Alphabet{T}}} where T" href="#BioSequenceMappings.Alignment-Union{Tuple{T}, Tuple{AbstractMatrix{T}, Alphabet{T}}} where T"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">Alignment(data::AbstractMatrix, alphabet; kwargs...)</code></pre><p><code>data</code> is a matrix of integers. <code>alphabet</code> can be either     - an <code>Alphabet</code>     - <code>nothing</code>     - something to build an alphabet from: the constructor <code>Alphabet</code> will be called.     (<em>e.g.</em> a symbol like <code>:aa</code>, a string, ...)</p><p>If the types of <code>alphabet</code> and <code>data</code> mismatch, <code>data</code> is converted.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alignment.jl#L93-L104">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alphabet" href="#BioSequenceMappings.Alphabet"><code>BioSequenceMappings.Alphabet</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Alphabet{T}
    string::String
    char_to_index::Dict{Char, T}
    index_to_char::Dict{T, Char}
end</code></pre><p><code>Alphabet</code> can be constructed</p><ul><li>from a <code>String</code> and an optional type: <code>Alphabet(string::String[, T&lt;:Integer])</code></li><li>from a mapping <code>Dict{Char, T}</code> where <code>T&lt;:Integer</code>: <code>Alphabet(mapping)</code></li><li>from a <code>Symbol</code>, using default alphabets</li><li>from an integer, using default alphabets (see <code>default_alphabets</code>).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alphabet.jl#L24-L36">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.compute_mapping-Union{Tuple{AbstractString}, Tuple{T}, Tuple{AbstractString, Type{T}}} where T&lt;:Integer" href="#BioSequenceMappings.compute_mapping-Union{Tuple{AbstractString}, Tuple{T}, Tuple{AbstractString, Type{T}}} where T&lt;:Integer"><code>BioSequenceMappings.compute_mapping</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">compute_mapping(s::AbstractString)</code></pre><p>Return a <code>Dict{Int, Char}</code>: <code>Dict(i =&gt; c for (i,c) in enumerate(s))</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alphabet.jl#L6-L10">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.compute_weights" href="#BioSequenceMappings.compute_weights"><code>BioSequenceMappings.compute_weights</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">compute_weights(X::AbstractAlignment, θ = 0.2; normalize = true)</code></pre><p>Compute phylogenetic correction weights for sequences of <code>X</code>.     The weight sequence <code>S</code> is <code>1/N</code>, where <code>N</code> is the number of sequences in <code>X</code> at     hamming distance less than <code>H</code> from <code>S</code> (including <code>S</code> itself).     The threshold <code>H</code> is <code>floor(θ⋅L)</code> where <code>L</code> is the sequence length.</p><p>The return value is a tuple <code>(weights, Meff)</code>, where <code>Meff</code> is the sum of weights (pre-normalization). If <code>normalize</code>, weights are normalized to sum to one. .</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/weights.jl#L1-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.compute_weights!-Tuple{Any, Any}" href="#BioSequenceMappings.compute_weights!-Tuple{Any, Any}"><code>BioSequenceMappings.compute_weights!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">compute_weights!(X, θ; kwargs...)</code></pre><p>Compute and set weights for <code>X</code>. See <code>compute_weights</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/weights.jl#L23-L27">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.default_alphabet-Union{Tuple{Integer}, Tuple{T}, Tuple{Integer, Type{T}}} where T&lt;:Integer" href="#BioSequenceMappings.default_alphabet-Union{Tuple{Integer}, Tuple{T}, Tuple{Integer, Type{T}}} where T&lt;:Integer"><code>BioSequenceMappings.default_alphabet</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">default_alphabet(q::Int, T::Type)</code></pre><ul><li>if <code>q==21</code>, amino acids</li><li>if <code>q==5</code>, nucleotides</li><li>if <code>q==4</code>, nucleotides without gaps</li><li>if <code>q==2</code>, binary (0, 1)</li><li>else, if <code>q&lt;21</code>, return the restriction of amino acids to the first q sites</li><li>if <code>q&gt;21</code>, fails</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alphabet.jl#L169-L178">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.pairwise_hamming-Tuple{Alignment, Alignment}" href="#BioSequenceMappings.pairwise_hamming-Tuple{Alignment, Alignment}"><code>BioSequenceMappings.pairwise_hamming</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pairwise_hamming(X, Y; step=1, step_left, step_right)</code></pre><p>Return matrix of all hamming distances between sequences of <code>X</code> and <code>Y</code>. Only consider sequences every <code>step</code>. <code>step_left</code> and <code>step_right</code> can be used to skip sequence either in <code>X</code> or in <code>Y</code>:</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/misc.jl#L17-L23">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.pairwise_hamming-Tuple{Alignment}" href="#BioSequenceMappings.pairwise_hamming-Tuple{Alignment}"><code>BioSequenceMappings.pairwise_hamming</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pairwise_hamming(X; step, as_vec=true)</code></pre><p>Vector of pairwise hamming distances of sequences in <code>X</code>, ordered as <code>[H(1,2), H(1,3), ..., H(M-1, M)]</code> with <code>H</code> standing for hamming distance. If <code>as_vec=false</code>, will return a <code>Matrix</code> instead.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/misc.jl#L38-L44">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.read_fasta-Tuple{AbstractString}" href="#BioSequenceMappings.read_fasta-Tuple{AbstractString}"><code>BioSequenceMappings.read_fasta</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">read_fasta(fastafile::AbstractString; alphabet = :auto, kwargs...)
read_fasta(
    fastafile::AbstractString, alphabet;
    weights = false, theta = 0.2, verbose = false,
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/IO.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.site_specific_frequencies-Tuple{Any, Vararg{Any}}" href="#BioSequenceMappings.site_specific_frequencies-Tuple{Any, Vararg{Any}}"><code>BioSequenceMappings.site_specific_frequencies</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">site_specific_frequencies(X::AbstractAlignment, weights=nothing)</code></pre><p>Return the site specific frequencies of <code>X</code>. If <code>as_vec</code> (default), the result is a vector of length <code>Lxq</code>. Otherwise, it is a matrix of <code>q</code> rows and <code>L</code> columns</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/statistics.jl#L25-L31">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.subsample-Tuple{BioSequenceMappings.AbstractAlignment, Int64}" href="#BioSequenceMappings.subsample-Tuple{BioSequenceMappings.AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">subsample(X::AbstractAlignment, indices)</code></pre><p>Return an <code>Alignment</code> containing only the sequences of <code>X</code> at <code>indices</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alignment.jl#L274-L278">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.subsample_random-Tuple{BioSequenceMappings.AbstractAlignment, Int64}" href="#BioSequenceMappings.subsample_random-Tuple{BioSequenceMappings.AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample_random</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">subsample_random(X::AbstractAlignment, m::Int)</code></pre><p>Return an <code>Alignment</code> with <code>m</code> sequences taking randomly from <code>X</code>. Sampling is done without replacement, meaning the <code>m</code> sequences are all at different positions in <code>X</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/e1468fd3d8dd3faa7bb1abff37c8af7d8ebcc4f1/src/alignment.jl#L289-L295">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Wednesday 22 May 2024 07:41">Wednesday 22 May 2024</span>. Using Julia version 1.10.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
