<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Alignments · BioSequenceMappings.jl</title><meta name="title" content="Alignments · BioSequenceMappings.jl"/><meta property="og:title" content="Alignments · BioSequenceMappings.jl"/><meta property="twitter:title" content="Alignments · BioSequenceMappings.jl"/><meta name="description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:description" content="Documentation for BioSequenceMappings.jl."/><meta property="twitter:description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/alignments/"/><meta property="twitter:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/alignments/"/><link rel="canonical" href="https://pierrebarrat.github.io/BioSequenceMappings.jl/alignments/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioSequenceMappings.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Quickstart</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../alphabets/">Alphabets</a></li><li class="is-active"><a class="tocitem" href>Alignments</a><ul class="internal"><li><a class="tocitem" href="#Reading-and-writing"><span>Reading &amp; writing</span></a></li><li><a class="tocitem" href="#Accessing-and-iterating"><span>Accessing &amp; iterating</span></a></li><li><a class="tocitem" href="#Finding-sequences"><span>Finding sequences</span></a></li><li><a class="tocitem" href="#Creating-subalignments"><span>Creating subalignments</span></a></li><li><a class="tocitem" href="#OneHotAlignment"><span>OneHotAlignment</span></a></li></ul></li><li><a class="tocitem" href="../utilities/">Utilities</a></li></ul></li><li><a class="tocitem" href="../reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Alignments</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Alignments</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/master/docs/src/alignments.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="The-Alignment-type"><a class="docs-heading-anchor" href="#The-Alignment-type">The <code>Alignment</code> type</a><a id="The-Alignment-type-1"></a><a class="docs-heading-anchor-permalink" href="#The-Alignment-type" title="Permalink"></a></h1><p>An <a href="../reference/#BioSequenceMappings.Alignment"><code>Alignment</code></a> essentially contains a set of aligned biological sequences mapped to integers.  It is a subtype of the more general <code>AbstractAlignment</code>.  The precise type is of the form <code>Alignment{A,T}</code> where <code>A</code> is the type of biological symbols (see <a href="../alphabets/#The-Alphabet-type">The <code>Alphabet</code> type</a>) and <code>T</code> the type of integers.  Its fields are simple and self-explanatory:</p><ul><li><code>data::Matrix{T}</code>: the mapped sequences. Each column of the matrix is one sequence. </li><li><code>alphabet::Alphabet{A,T}</code>: the alphabet defining the mapping from symbols to integers.</li><li><code>weights::Vector{Float64}</code>: weights associated with each sequence. DCA-like methods assign phylogenetic weights to sequences when infering models. </li><li><code>names::Vector{String}</code>: names of the sequences, corresponding to the description string if the alignment was built from a FASTA file. </li></ul><p>The dimensions of an alignment can be obtained using <code>size</code>: length <span>$\cross$</span> number of sequences</p><h2 id="Reading-and-writing"><a class="docs-heading-anchor" href="#Reading-and-writing">Reading &amp; writing</a><a id="Reading-and-writing-1"></a><a class="docs-heading-anchor-permalink" href="#Reading-and-writing" title="Permalink"></a></h2><p>For now, the only file format that this package interacts with is FASTA.  Reading an alignment is simple: </p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A = read_fasta(&quot;../../example/PF00014.fasta&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=100 sequences of length L=53 - Shown as `MxL` matrix
100×53 adjoint(::Matrix{Int64}) with eltype Int64:
  1  3  10  10  14  14  17   5   7  19  …  15  15   4  3   9  17  18  3   1
  1  3  15  15  14  11   8   8   7  13      6   5   4  1   1   1   1  1   1
 19  3   8  11  14  16   5   5   7   5     16  10  15  3  16  10   9  3   2
 18  3  17  11  17  14  17  14   7  18     16   2   5  3   5  13  21  3   1
 19  3   2   6  14  12   5  10   7  14     10   5  10  3   5  10   6  3  10
 11  3  11  15  17  10   4   9   7  18  …   5   5   2  3   5  16   2  3   1
  2  3  13  11  14   9  19  16   7  14      5  10   5  3  10   5  21  3   1
 19  3  17  11  14   2  19  15   7  14     16   5   5  3   5   5  17  3   1
  6  3   2  11  10   2   4   4   7  14     11   5   5  3  10   5  10  3   1
  1  3  11   2   5  19   4  11  16  19     21   5   5  3  10   2  17  3  15
  ⋮                  ⋮                  ⋱       ⋮                  ⋮
  9  3  12   5  10  14   5  10   7  14     17  18   5  3  17  10  18  3   1
  2  3   7  11  14  10   5  18   7  14      5  17   5  3  10   5  18  3  15
  1  3  13  11  14  17  19  19   7   9     10   5  17  3   5  16  15  3   1
  9  3  17   6  14  14   4  20   7  17      6   5   9  3  15  10   2  3   5
  2  3   5  11  14  19  13   8   7  16  …  21   5   5  3   9   7  12  3   1
  1  3  10  11  16  11   4  15   7  17      5   5   5  3  10  10  18  3   1
 14  3   5  15  14  12   4   2   7  14      5   2  17  3   7  21   8  3  10
  1  3  21  19   7  16  13   5   7  21     15   5  17  3   5   2  16  3   1
  9  3  12  11  14  10   4  18   7  18     15   5   2  3   5   2  19  3   2</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; size(A) # length and number of sequences</code><code class="nohighlight hljs ansi" style="display:block;">(53, 100)</code></pre><p>When reading from FASTA, the choice of the alphabet is made by reading the first five sequences, and comparing the observed characters with the list of default alphabets (see <a href="../alphabets/#The-Alphabet-type">The <code>Alphabet</code> type</a>).  If they fit one of the defaults, it will be used.  Otherwise, an alphabet will be created <em>ad hoc</em>: </p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A = read_fasta(&quot;../../example/strange_characters.fasta&quot;); # warning produced because no default alphabet was found</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr33"><span class="sgr1">┌ Warning: </span></span>Could not find a default alphabet for characters [&#39;!&#39;, &#39;-&#39;, &#39;/&#39;, &#39;0&#39;, &#39;9&#39;, &#39;@&#39;, &#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;]
<span class="sgr33"><span class="sgr1">│ </span></span> Using Alphabet{Char,Int64}: [&#39;!&#39;, &#39;-&#39;, &#39;/&#39;, &#39;0&#39;, &#39;9&#39;, &#39;@&#39;, &#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;]
<span class="sgr33"><span class="sgr1">└ </span></span><span class="sgr90">@ BioSequenceMappings ~/work/BioSequenceMappings.jl/BioSequenceMappings.jl/src/IO.jl:62</span></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A.alphabet |&gt; symbols |&gt; prod</code><code class="nohighlight hljs ansi" style="display:block;">&quot;!-/09@ACGT&quot;</code></pre><p>Writing to a FASTA file is just as easy: </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; write(&quot;new_fasta_file.fasta&quot;, A) # or...</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; open(&quot;new_fasta_file.fasta&quot;, &quot;w&quot;) do io
       	write(io, A)
       end</code><code class="nohighlight hljs ansi" style="display:block;"></code></pre><h2 id="Accessing-and-iterating"><a class="docs-heading-anchor" href="#Accessing-and-iterating">Accessing &amp; iterating</a><a id="Accessing-and-iterating-1"></a><a class="docs-heading-anchor-permalink" href="#Accessing-and-iterating" title="Permalink"></a></h2><p>Sequences can be accessed by indexing.  Indexing using a range will return a view in the underlying <code>data</code> matrix. </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; A[1] # the first sequence of the alignment</code><code class="nohighlight hljs ansi" style="display:block;">53-element view(::Matrix{Int64}, :, 1) with eltype Int64:
  1
  3
 10
 10
 14
 14
 17
  5
  7
 19
  ⋮
 15
 15
  4
  3
  9
 17
 18
  3
  1</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; size(A[1:5]) # 5 sequences of length L</code><code class="nohighlight hljs ansi" style="display:block;">(53, 5)</code></pre><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>When indexing or when iterating, the return value is a <em>reference</em> to the sequence and not a copy.  For this reason, if you modify the sequence, the alignment itself will be modified</p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; s = A[1]&#39;</code><code class="nohighlight hljs ansi" style="display:block;">1×53 adjoint(view(::Matrix{Int64}, :, 1)) with eltype Int64:
 1  3  10  10  14  14  17  5  7  19  3  …  17  15  15  4  3  9  17  18  3  1</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; s[1] = 21</code><code class="nohighlight hljs ansi" style="display:block;">21</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A[1]&#39; # the first element has been modified</code><code class="nohighlight hljs ansi" style="display:block;">1×53 adjoint(view(::Matrix{Int64}, :, 1)) with eltype Int64:
 21  3  10  10  14  14  17  5  7  19  3  …  17  15  15  4  3  9  17  18  3  1</code></pre></div></div><p>The fact that calls like <code>A[1:5]</code> return a matrix-like object can be inconvenient.  To iterate over sequences as vectors of integers, one can use the <a href="../reference/#BioSequenceMappings.eachsequence-Tuple{AbstractAlignment, Any}"><code>eachsequence</code></a> function.  It takes the alignment as a first argument and optionally indices, and return an iterator. </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; for s in eachsequence(A)
       	# s: vector of integers
       end</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; map(length, eachsequence(A, 1:5:16)) # length of sequences 1, 6, 11. 16</code><code class="nohighlight hljs ansi" style="display:block;">4-element Vector{Int64}:
 53
 53
 53
 53</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; collect(eachsequence(A; skip=25)) # collect as list of vectors, taking one every 25 sequences</code><code class="nohighlight hljs ansi" style="display:block;">4-element Vector{SubArray{Int64, 1, Matrix{Int64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}}:
 [1, 3, 10, 10, 14, 14, 17, 5, 7, 19  …  17, 15, 15, 4, 3, 9, 17, 18, 3, 1]
 [1, 3, 17, 5, 16, 6, 4, 2, 2, 14  …  18, 5, 18, 4, 3, 11, 13, 18, 3, 1]
 [19, 3, 11, 11, 15, 10, 4, 18, 7, 14  …  17, 10, 15, 5, 3, 5, 18, 4, 3, 1]
 [1, 3, 19, 7, 19, 17, 11, 14, 7, 14  …  17, 16, 5, 2, 3, 5, 4, 16, 3, 2]</code></pre><p>If the name of the sequence is needed, the iterator <a href="../reference/#BioSequenceMappings.named_sequences-Tuple{AbstractAlignment, Any}"><code>named_sequences</code></a> can be used instead, taking the same arguments and iterating over tuples of the form <code>(name, sequence)</code>. </p><h2 id="Finding-sequences"><a class="docs-heading-anchor" href="#Finding-sequences">Finding sequences</a><a id="Finding-sequences-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-sequences" title="Permalink"></a></h2><p>The package provides <a href="../reference/#BioSequenceMappings.find_sequence-Tuple{AbstractString, AbstractAlignment}"><code>find_sequence</code></a> to find sequences by name in an alignment, and <a href="../reference/#BioSequenceMappings.match_sequences-Tuple{Any, AbstractAlignment}"><code>match_sequences</code></a> to match all sequences with a particular name pattern. </p><pre><code class="language-julia-repl hljs">julia&gt; A = read_fasta(joinpath(example_dir, &quot;toy_fasta_dna.fasta&quot;));

julia&gt; n = A.names[1] # name of the first sequence in the alignment
&quot;sequence_1&quot;

julia&gt; find_sequence(n, A) # index and sequence
(1, [5, 3, 4, 2, 4, 1, 5, 1, 5, 5])

julia&gt; indices, sequences = match_sequences(r&quot;sequence_[0-9]&quot;, A); # using a regex 

julia&gt; indices
5-element Vector{Int64}:
 1
 2
 3
 4
 5</code></pre><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p><code>find_sequence</code> searches for an exact match.  It is based on <code>findfirst</code>, and will thus return the first match it finds.  If nothing is found, then the result is <code>nothing</code>.  On the other hand, <code>match_sequences</code> is based on <code>occursin</code> and <code>findall</code>.  The returned sequences are references to original objects, as when indexing. </p></div></div><h2 id="Creating-subalignments"><a class="docs-heading-anchor" href="#Creating-subalignments">Creating subalignments</a><a id="Creating-subalignments-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-subalignments" title="Permalink"></a></h2><p>The call <code>subsample(A, indices)</code> will create an <code>AbstractAlignment</code> of the same type as <code>A</code> by taking only the sequences at <code>indices</code>.  It uses the same alphabet as <code>A</code> and copies over the names of the sequences.  Note that <a href="../reference/#BioSequenceMappings.subsample-Tuple{AbstractAlignment, Int64}"><code>subsample</code></a> <em>copies</em> the underlying data, creating a completely independent object.</p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A = read_fasta(&quot;../../example/toy_fasta_dna.fasta&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=5 sequences of length L=10 - Shown as `MxL` matrix
5×10 adjoint(::Matrix{Int64}) with eltype Int64:
 5  3  4  2  4  1  5  1  5  5
 2  4  4  1  4  2  3  2  2  3
 4  3  5  3  3  5  5  3  5  1
 3  3  5  2  2  1  5  3  1  2
 2  3  2  5  3  2  4  3  2  4</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A[1][1] # first character of the first sequence</code><code class="nohighlight hljs ansi" style="display:block;">5</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; B = subsample(A, 1:2:5)</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=3 sequences of length L=10 - Shown as `MxL` matrix
3×10 adjoint(::Matrix{Int64}) with eltype Int64:
 5  3  4  2  4  1  5  1  5  5
 4  3  5  3  3  5  5  3  5  1
 2  3  2  5  3  2  4  3  2  4</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; B[1][1] = 5</code><code class="nohighlight hljs ansi" style="display:block;">5</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A[1][1] # A remains unchanged</code><code class="nohighlight hljs ansi" style="display:block;">5</code></pre><p>With <a href="../reference/#BioSequenceMappings.subsample_random-Tuple{AbstractAlignment, Int64}"><code>subsample_random</code></a>, it is also possible to create a random subalignment by picking sequences from the original one.  For now, this is only possible without replacement, <em>i.e.</em> the same sequence cannot be picked twice.  To just pick one sequence at random without creating a new alignment object, just call <code>rand</code>. </p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; subsample_random(A, 3) # new alignment using three random sequences from A</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=3 sequences of length L=10 - Shown as `MxL` matrix
3×10 adjoint(::Matrix{Int64}) with eltype Int64:
 3  3  5  2  2  1  5  3  1  2
 2  4  4  1  4  2  3  2  2  3
 5  3  4  2  4  1  5  1  5  5</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; subsample_random(A, 12) # sampling without replacement: this will error since size(A, 1) &lt; 12</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: AssertionError: Cannot take 12 different sequences from alignment of size 5</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; rand(A) # one random sequence from A (returns a view)</code><code class="nohighlight hljs ansi" style="display:block;">10-element view(::Matrix{Int64}, :, 4) with eltype Int64:
 3
 3
 5
 2
 2
 1
 5
 3
 1
 2</code></pre><h2 id="OneHotAlignment"><a class="docs-heading-anchor" href="#OneHotAlignment">OneHotAlignment</a><a id="OneHotAlignment-1"></a><a class="docs-heading-anchor-permalink" href="#OneHotAlignment" title="Permalink"></a></h2><p>TBA</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../alphabets/">« Alphabets</a><a class="docs-footer-nextpage" href="../utilities/">Utilities »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.7.0 on <span class="colophon-date" title="Monday 30 September 2024 16:22">Monday 30 September 2024</span>. Using Julia version 1.10.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
