<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Utilities · BioSequenceMappings.jl</title><meta name="title" content="Utilities · BioSequenceMappings.jl"/><meta property="og:title" content="Utilities · BioSequenceMappings.jl"/><meta property="twitter:title" content="Utilities · BioSequenceMappings.jl"/><meta name="description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:description" content="Documentation for BioSequenceMappings.jl."/><meta property="twitter:description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/utilities/"/><meta property="twitter:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/utilities/"/><link rel="canonical" href="https://pierrebarrat.github.io/BioSequenceMappings.jl/utilities/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioSequenceMappings.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Quickstart</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../alphabets/">Alphabets</a></li><li><a class="tocitem" href="../alignments/">Alignments</a></li><li class="is-active"><a class="tocitem" href>Utilities</a><ul class="internal"><li><a class="tocitem" href="#Hamming-distance"><span>Hamming distance</span></a></li><li><a class="tocitem" href="#Alignment-statistics"><span>Alignment statistics</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../reference/">Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Utilities</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Utilities</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/master/docs/src/utilities.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Utilities"><a class="docs-heading-anchor" href="#Utilities">Utilities</a><a id="Utilities-1"></a><a class="docs-heading-anchor-permalink" href="#Utilities" title="Permalink"></a></h1><h2 id="Hamming-distance"><a class="docs-heading-anchor" href="#Hamming-distance">Hamming distance</a><a id="Hamming-distance-1"></a><a class="docs-heading-anchor-permalink" href="#Hamming-distance" title="Permalink"></a></h2><p><a href="../reference/#BioSequenceMappings.hamming-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>hamming(x, y)</code></a> returns the normalized Hamming distance between two integer vectors. </p><h2 id="Alignment-statistics"><a class="docs-heading-anchor" href="#Alignment-statistics">Alignment statistics</a><a id="Alignment-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Alignment-statistics" title="Permalink"></a></h2><h3 id="Pairwise-Hamming-distance"><a class="docs-heading-anchor" href="#Pairwise-Hamming-distance">Pairwise Hamming distance</a><a id="Pairwise-Hamming-distance-1"></a><a class="docs-heading-anchor-permalink" href="#Pairwise-Hamming-distance" title="Permalink"></a></h3><p><a href="../reference/#BioSequenceMappings.pairwise_hamming-Tuple{AbstractAlignment, AbstractAlignment}"><code>pairwise_hamming(A)</code></a> returns a vector containing all Hamming distances between pairs of sequences in <code>A</code>.  For large alignments, it is often practical to consider only a subset of Hamming distances: the <code>step=n</code> keyword can be used to only consider every nth sequence.  The function is also adapted to two alignments: [<code>pairwise_hamming(A,B)</code>] will consider all pairs of sequences with one member in <code>A</code> and the other one in <code>B</code>. </p><h3 id="Statistics"><a class="docs-heading-anchor" href="#Statistics">Statistics</a><a id="Statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Statistics" title="Permalink"></a></h3><ul><li>the profile of the alignment: <a href="../reference/#BioSequenceMappings.site_specific_frequencies"><code>site_specific_frequencies(A)</code></a> returns a <code>q x L</code> matrix where <code>q</code> is the size of the alphabet and <code>L</code> the length of sequences, with element <code>(a, i)</code> being the fraction of sequences in which character <code>a</code> was found at position <code>i</code>. </li><li>the pairwise frequencies: <a href="../reference/#BioSequenceMappings.pairwise_frequencies"><code>pairwise_frequencies(A)</code></a></li><li>the pairwise correlations: <a href="../reference/#BioSequenceMappings.pairwise_correlations"><code>pairwise_correlations(A)</code></a></li></ul><p>All these functions accept a vector of weights as a second argument, and will by default use <code>A.weights</code> if it is not provided. </p><h3 id="Weights"><a class="docs-heading-anchor" href="#Weights">Weights</a><a id="Weights-1"></a><a class="docs-heading-anchor-permalink" href="#Weights" title="Permalink"></a></h3><p>In DCA-like algorithms, it is customary to weight sequences by their degree of phylogenetic relations.  Typically, the weight of a sequence is inversely proportional to the number of other sequences with a Hamming distance smaller than some threshold.  For an alignment <code>X</code>, computing the weights is done using <a href="../reference/#BioSequenceMappings.compute_weights"><code>compute_weights</code></a>: </p><pre><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A = read_fasta(&quot;../../example/toy_fasta_dna.fasta&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">Alignment of M=5 sequences of length L=10 - Shown as `MxL` matrix
5×10 adjoint(::Matrix{Int64}) with eltype Int64:
 5  3  4  2  4  1  5  1  5  5
 2  4  4  1  4  2  3  2  2  3
 4  3  5  3  3  5  5  3  5  1
 3  3  5  2  2  1  5  3  1  2
 2  3  2  5  3  2  4  3  2  4</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; compute_weights(A) # compute and return the weight vector, as well as the effective number of sequences</code><code class="nohighlight hljs ansi" style="display:block;">([0.2, 0.2, 0.2, 0.2, 0.2], 5.0)</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; compute_weights!(A); # same, but sets the weights field in A</code><code class="nohighlight hljs ansi" style="display:block;"></code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; A.weights</code><code class="nohighlight hljs ansi" style="display:block;">5-element Vector{Float64}:
 0.2
 0.2
 0.2
 0.2
 0.2</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../alignments/">« Alignments</a><a class="docs-footer-nextpage" href="../reference/">Reference »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.7.0 on <span class="colophon-date" title="Monday 30 September 2024 16:16">Monday 30 September 2024</span>. Using Julia version 1.10.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
