<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Reference · BioSequenceMappings.jl</title><meta name="title" content="Reference · BioSequenceMappings.jl"/><meta property="og:title" content="Reference · BioSequenceMappings.jl"/><meta property="twitter:title" content="Reference · BioSequenceMappings.jl"/><meta name="description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:description" content="Documentation for BioSequenceMappings.jl."/><meta property="twitter:description" content="Documentation for BioSequenceMappings.jl."/><meta property="og:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/reference/"/><meta property="twitter:url" content="https://pierrebarrat.github.io/BioSequenceMappings.jl/reference/"/><link rel="canonical" href="https://pierrebarrat.github.io/BioSequenceMappings.jl/reference/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BioSequenceMappings.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Quickstart</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../alphabets/">Alphabets</a></li><li><a class="tocitem" href="../alignments/">Alignments</a></li><li><a class="tocitem" href="../utilities/">Utilities</a></li></ul></li><li class="is-active"><a class="tocitem" href>Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/master/docs/src/reference.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><p>Documentation for <a href="https://github.com/PierreBarrat/BioSequenceMappings.jl">BioSequenceMappings</a>.</p><ul><li><a href="#BioSequenceMappings.Alignment"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#BioSequenceMappings.Alignment-Union{Tuple{T}, Tuple{A}, Tuple{AbstractMatrix{T}, Alphabet{A, T}}} where {A, T}"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#BioSequenceMappings.Alignment-Union{Tuple{AbstractMatrix{T}}, Tuple{T}} where T&lt;:Integer"><code>BioSequenceMappings.Alignment</code></a></li><li><a href="#BioSequenceMappings.Alphabet"><code>BioSequenceMappings.Alphabet</code></a></li><li><a href="#BioSequenceMappings.compute_weights"><code>BioSequenceMappings.compute_weights</code></a></li><li><a href="#BioSequenceMappings.compute_weights!"><code>BioSequenceMappings.compute_weights!</code></a></li><li><a href="#BioSequenceMappings.default_alphabet-Union{Tuple{Integer}, Tuple{T}, Tuple{Integer, Type{T}}} where T&lt;:Integer"><code>BioSequenceMappings.default_alphabet</code></a></li><li><a href="#BioSequenceMappings.eachsequence-Tuple{AbstractAlignment, Any}"><code>BioSequenceMappings.eachsequence</code></a></li><li><a href="#BioSequenceMappings.find_sequence-Tuple{AbstractString, AbstractAlignment}"><code>BioSequenceMappings.find_sequence</code></a></li><li><a href="#BioSequenceMappings.hamming-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>BioSequenceMappings.hamming</code></a></li><li><a href="#BioSequenceMappings.match_sequences-Tuple{Any, AbstractAlignment}"><code>BioSequenceMappings.match_sequences</code></a></li><li><a href="#BioSequenceMappings.named_sequences-Tuple{AbstractAlignment, Any}"><code>BioSequenceMappings.named_sequences</code></a></li><li><a href="#BioSequenceMappings.pairwise_correlations"><code>BioSequenceMappings.pairwise_correlations</code></a></li><li><a href="#BioSequenceMappings.pairwise_frequencies"><code>BioSequenceMappings.pairwise_frequencies</code></a></li><li><a href="#BioSequenceMappings.pairwise_hamming-Tuple{AbstractAlignment, AbstractAlignment}"><code>BioSequenceMappings.pairwise_hamming</code></a></li><li><a href="#BioSequenceMappings.read_fasta-Tuple{AbstractString}"><code>BioSequenceMappings.read_fasta</code></a></li><li><a href="#BioSequenceMappings.site_specific_frequencies"><code>BioSequenceMappings.site_specific_frequencies</code></a></li><li><a href="#BioSequenceMappings.subsample-Tuple{AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample</code></a></li><li><a href="#BioSequenceMappings.subsample_random-Tuple{AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample_random</code></a></li><li><a href="#BioSequenceMappings.symbols-Tuple{Alphabet}"><code>BioSequenceMappings.symbols</code></a></li><li><a href="#BioSequenceMappings.translate-Tuple{Integer, Alphabet, Alphabet}"><code>BioSequenceMappings.translate</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alignment" href="#BioSequenceMappings.Alignment"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mutable struct Alignment{A,T} where {A, T&lt;:Integer}</code></pre><pre><code class="nohighlight hljs">    data::Matrix{T}
    alphabet::Union{Nothing, Alphabet{A,T}}
    weights::Vector{Float64} = ones(size(dat,1))/size(dat,1) # phylogenetic weights of sequences
    names::Vector{String} = fill(&quot;&quot;, size(dat, 1))</code></pre><p>Biological sequences as vectors of type <code>T&lt;:Integer</code>. <code>data</code> stores sequences in <em>columns</em>: <code>size(dat)</code> returns a tuple <code>(L, M)</code> with <code>L</code> the length and <code>M</code> the number of sequences. When displayed, shows <code>data</code> as an <code>MxL</code> matrix to match with traditional alignments.</p><p><code>alphabet{A,T}</code> represents the mapping between integers in <code>data</code> and biological symbols of type <code>A</code> (nucleotides, amino acids...). If <code>nothing</code>, the alignment cannot be mapped to biological sequences.</p><p><code>weights</code> represent phylogenetic weights, and are initialized to <code>1/M</code>. They must sum to 1. <code>names</code> are the label of sequences, and are expected to be in the same order as the columns of <code>data</code>. They do not have to be unique, and can be ignored</p><p><strong>Important</strong>: When built from a matrix, assumes that the sequences are stored in <em>columns</em>.</p><p><strong>Methods</strong></p><ul><li><code>getindex(X::Alignment, i)</code> returns a matrix/vector <code>X.data[:, i]</code>.</li><li><code>for s in X::Alignment</code> iterates over sequences.</li><li><code>eachsequence(X::Alignment)</code> returns an iterator over sequences (<code>Vector{Int}</code>).</li><li><code>eachsequence_weighted(X::Alignment)</code> returns an iterator over sequences and weights as tuples</li><li><code>subaln(X::Alignment, idx)</code> constructs the subaln defined by index <code>idx</code>.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L7-L38">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alignment-Union{Tuple{AbstractMatrix{T}}, Tuple{T}} where T&lt;:Integer" href="#BioSequenceMappings.Alignment-Union{Tuple{AbstractMatrix{T}}, Tuple{T}} where T&lt;:Integer"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Alignment(data::AbstractMatrix{T}; alphabet = :auto, kwargs...)</code></pre><p>Keyword argument <code>alphabet</code> can be <code>:auto</code>, <code>:none</code>/<code>nothing</code>, or an input to the constructor <code>Alphabet</code>. Other keyword arguments are passed to the default constructor of <code>Alignment</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L141-L147">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alignment-Union{Tuple{T}, Tuple{A}, Tuple{AbstractMatrix{T}, Alphabet{A, T}}} where {A, T}" href="#BioSequenceMappings.Alignment-Union{Tuple{T}, Tuple{A}, Tuple{AbstractMatrix{T}, Alphabet{A, T}}} where {A, T}"><code>BioSequenceMappings.Alignment</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Alignment(data::AbstractMatrix, alphabet; kwargs...)</code></pre><p><code>data</code> is a matrix of integers, with sequences stored in columns. <code>alphabet</code> can be either</p><ul><li>an <code>Alphabet</code></li><li><code>nothing</code>: no conversion from integers to biological symbols.</li><li>something to build an alphabet from (<em>e.g.</em> a symbol like <code>:aa</code>, a string, ...).   The constructor <code>Alphabet</code> will be called like so: <code>Alphabet(alphabet)</code>.</li></ul><p>If the types of <code>alphabet</code> and <code>data</code> mismatch, <code>data</code> is converted.</p><p><code>data</code> can also have the following shape:</p><ul><li>vector of integer vectors, <em>e.g.</em> [[1,2], [3,4]]: each element is considered as a sequence</li><li>vector of integers: single sequence alignment</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L103-L118">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.Alphabet" href="#BioSequenceMappings.Alphabet"><code>BioSequenceMappings.Alphabet</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">struct Alphabet{A,I}
    characters::Vector{A}
    char_to_index::Dict{A, I}
    index_to_char::Dict{I, A}
    default_char = nothing
    default_index
end</code></pre><p>Structure allowing the mapping from biological symbols of type <code>A</code> to integers of type <code>I</code>.     The typical use case would be <code>Alphabet{Char, Int}</code>. <code>Alphabet</code> can be constructed</p><ul><li>from a <code>Vector</code> of symbols and an optional type <code>I</code>, <em>e.g.</em> <code>Alphabet([&#39;A&#39;,&#39;C&#39;,&#39;G&#39;,&#39;T&#39;], UInt8)::Alphabet{Char, UInt8}</code></li><li>from a <code>String</code> and an optional type, <em>e.g.</em> <code>Alphabet(&quot;ACGT&quot;)</code></li><li>from a mapping <code>Dict{A, I}</code> where <code>I&lt;:Integer</code>: <code>Alphabet(Dict(&#39;A&#39;=&gt;1, &#39;C&#39;=&gt;2))</code></li><li>from a <code>Symbol</code>, using default alphabets, <em>e.g.</em> <code>Alphabet(:nt)</code></li><li>from an integer, using default alphabets (see <code>?default_alphabets</code>).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alphabet.jl#L24-L41">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.compute_weights" href="#BioSequenceMappings.compute_weights"><code>BioSequenceMappings.compute_weights</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">compute_weights(X::AbstractAlignment, θ = 0.2; normalize = true)</code></pre><p>Compute phylogenetic correction weights for sequences of <code>X</code>.     The weight sequence <code>S</code> is <code>1/N</code>, where <code>N</code> is the number of sequences in <code>X</code> at     hamming distance less than <code>H</code> from <code>S</code> (including <code>S</code> itself).     The threshold <code>H</code> is <code>floor(θ⋅L)</code> where <code>L</code> is the sequence length.</p><p>The return value is a tuple <code>(weights, Meff)</code>, where <code>Meff</code> is the sum of weights (pre-normalization). If <code>normalize</code>, weights are normalized to sum to one. .</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/weights.jl#L1-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.compute_weights!" href="#BioSequenceMappings.compute_weights!"><code>BioSequenceMappings.compute_weights!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">compute_weights!(X, θ; kwargs...)</code></pre><p>Compute and set weights for <code>X</code>. See <code>compute_weights</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/weights.jl#L23-L27">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.default_alphabet-Union{Tuple{Integer}, Tuple{T}, Tuple{Integer, Type{T}}} where T&lt;:Integer" href="#BioSequenceMappings.default_alphabet-Union{Tuple{Integer}, Tuple{T}, Tuple{Integer, Type{T}}} where T&lt;:Integer"><code>BioSequenceMappings.default_alphabet</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">default_alphabet(q::Int, T::Type)</code></pre><ul><li>if <code>q==2</code>, binary (0, 1)</li><li>if <code>3 &lt;= q &lt;= 4</code>, nucleotides without gaps</li><li>if <code>q==5</code>, nucleotides</li><li>if <code>5 &lt; q &lt;= 21</code>, amino acids</li><li>if <code>q&gt;21</code>, fails</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alphabet.jl#L245-L253">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.eachsequence-Tuple{AbstractAlignment, Any}" href="#BioSequenceMappings.eachsequence-Tuple{AbstractAlignment, Any}"><code>BioSequenceMappings.eachsequence</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">eachsequence(X::AbstractAlignment[, indices]; skip)</code></pre><p>Return an iterator over the sequences in <code>X</code>. If <code>indices</code> is specified, consider only sequences at the corresponding indices. Use the integer argument <code>skip</code> to return only one sequence every <code>skip</code> (~ <code>1:skip:end</code>).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L275-L281">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.find_sequence-Tuple{AbstractString, AbstractAlignment}" href="#BioSequenceMappings.find_sequence-Tuple{AbstractString, AbstractAlignment}"><code>BioSequenceMappings.find_sequence</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">find_sequence(label::AbstractString, aln::AbstractAlignment)</code></pre><p>Find sequence with name <code>label</code> in <code>aln</code>, and return <code>(index, sequence)</code>. Scales as the number of sequences. Return the first sequence that matches the label.</p><p>!!! Return a <em>view</em> of the sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L356-L364">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.hamming-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}" href="#BioSequenceMappings.hamming-Tuple{AbstractVector{&lt;:Integer}, AbstractVector{&lt;:Integer}}"><code>BioSequenceMappings.hamming</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">hamming(x, y; normalize=true, positions=nothing)</code></pre><p>Hamming distance between <code>Vectors</code> <code>x</code> and <code>y</code>. Only sites in vector <code>positions</code> will be considered.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/misc.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.match_sequences-Tuple{Any, AbstractAlignment}" href="#BioSequenceMappings.match_sequences-Tuple{Any, AbstractAlignment}"><code>BioSequenceMappings.match_sequences</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">match_sequences(pattern, aln::AbstractAlignment)</code></pre><p>Find sequences whose name matches <code>label</code> in <code>aln</code>, and return <code>(indices, sequences)</code>. Sequences are returned as columns.</p><p>!!! Return a <em>view</em> of the sequences.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L369-L376">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.named_sequences-Tuple{AbstractAlignment, Any}" href="#BioSequenceMappings.named_sequences-Tuple{AbstractAlignment, Any}"><code>BioSequenceMappings.named_sequences</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">named_sequences(X::AbstractAlignment; skip)</code></pre><p>Return an iterator of the form <code>(name, sequence)</code> over <code>X</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L294-L298">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.pairwise_correlations" href="#BioSequenceMappings.pairwise_correlations"><code>BioSequenceMappings.pairwise_correlations</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pairwise_correlations(X, w=X.weights; as_mat=false)</code></pre><p>Compute connected correlations: the difference between the pairwise frequencies and the product of the single site frequencies. See <code>?pairwise_frequencies</code> for the shape of the output.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/statistics.jl#L88-L94">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.pairwise_frequencies" href="#BioSequenceMappings.pairwise_frequencies"><code>BioSequenceMappings.pairwise_frequencies</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pairwise_frequencies(X::AbstractAlignment, w=X.weights; as_mat=false)</code></pre><p>Return a <code>q x q x L x L</code> tensor. The <code>(a, b, i, j)</code> element is the fraction of sequences for which we see <code>a</code> at position <code>i</code> and <code>b</code> at position <code>j</code>.</p><p>If <code>as_mat=true</code>, will return a <code>qL x qL</code> matrix, with <code>q x q</code> blocks representing correlations between two specific columns.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/statistics.jl#L68-L77">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.pairwise_hamming-Tuple{AbstractAlignment, AbstractAlignment}" href="#BioSequenceMappings.pairwise_hamming-Tuple{AbstractAlignment, AbstractAlignment}"><code>BioSequenceMappings.pairwise_hamming</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pairwise_hamming(X, Y; step=1, step_left, step_right, as_vec=true, kwargs...)
pairwise_hamming(X; kwargs...)</code></pre><p>Return all hamming distances between sequences of <code>X</code> and <code>Y</code>. In the second form, consider pairs of sequences in <code>X</code>.</p><p>Only consider sequences every <code>step</code>. <code>step_left</code> and <code>step_right</code> can be used to skip sequence either in <code>X</code> or in <code>Y</code>. This is useful for large alignment, as the number of computations grows with the product     of the size of the alignments</p><p>By default, the return value is a vector organized like <code>[H(1,2), H(1,3), ..., H(M-1, M)]</code> with <code>H</code> standing for hamming distance and <code>M</code> for the number of sequences. If a matrix is prefered, use <code>as_vec=false</code></p><p>Extra keyword arguments are passed to <code>hamming</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/misc.jl#L38-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.read_fasta-Tuple{AbstractString}" href="#BioSequenceMappings.read_fasta-Tuple{AbstractString}"><code>BioSequenceMappings.read_fasta</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_fasta(fastafile::AbstractString; alphabet = :auto, kwargs...)
read_fasta(
    fastafile::AbstractString, alphabet;
    weights = false, theta = 0.2, verbose = false,
)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/IO.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.site_specific_frequencies" href="#BioSequenceMappings.site_specific_frequencies"><code>BioSequenceMappings.site_specific_frequencies</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">site_specific_frequencies(X::AbstractAlignment[, weights=X.weights]; as_vec=false)</code></pre><p>Return the site specific frequencies of <code>X</code>. If <code>as_vec</code>, the result is a vector of length <code>Lxq</code>. Otherwise, it is a matrix of <code>q</code> rows and <code>L</code> columns (default).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/statistics.jl#L25-L31">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.subsample-Tuple{AbstractAlignment, Int64}" href="#BioSequenceMappings.subsample-Tuple{AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">subsample(X::AbstractAlignment, indices)</code></pre><p>Return an <code>Alignment</code> containing only the sequences of <code>X</code> at <code>indices</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L328-L332">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.subsample_random-Tuple{AbstractAlignment, Int64}" href="#BioSequenceMappings.subsample_random-Tuple{AbstractAlignment, Int64}"><code>BioSequenceMappings.subsample_random</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">subsample_random(X::AbstractAlignment, m::Int)</code></pre><p>Return an <code>Alignment</code> with <code>m</code> sequences taking randomly from <code>X</code>. Sampling is done without replacement, meaning the <code>m</code> sequences are all at different positions in <code>X</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alignment.jl#L343-L349">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.symbols-Tuple{Alphabet}" href="#BioSequenceMappings.symbols-Tuple{Alphabet}"><code>BioSequenceMappings.symbols</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">symbols(alphabet)</code></pre><p>Return the vector of symbols/characters used by <code>alphabet</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alphabet.jl#L362-L366">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="BioSequenceMappings.translate-Tuple{Integer, Alphabet, Alphabet}" href="#BioSequenceMappings.translate-Tuple{Integer, Alphabet, Alphabet}"><code>BioSequenceMappings.translate</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">translate(x, original_alphabet::Alphabet, new_alphabet::Alphabet)</code></pre><p>Return the translation in <code>new_alphabet</code> of an integer or a vector of integers <code>x</code> that is expressed in <code>original_alphabet</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/PierreBarrat/BioSequenceMappings.jl/blob/d39d61154861b4e6212592568513e2c22bcf4db1/src/alphabet.jl#L315-L320">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../utilities/">« Utilities</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.7.0 on <span class="colophon-date" title="Friday 25 October 2024 15:33">Friday 25 October 2024</span>. Using Julia version 1.11.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
